akka-stream/src/main/scala/akka/stream/impl/RetryFlowCoordinator.scala

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.annotation.InternalApi
+import akka.pattern.BackoffSupervisor
+import akka.stream.impl.RetryFlowCoordinator.InternalState
+import akka.stream.impl.RetryFlowCoordinator.RetryTimer
+import akka.stream.{ Attributes, BidiShape, Inlet, Outlet }
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler, TimerGraphStageLogic }
+
+import scala.collection.immutable
+import scala.concurrent.duration._
+import scala.util.Try
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object RetryFlowCoordinator {
+  final class InternalState(val numberOfRestarts: Int, val retryDeadline: Long)
+  case object RetryTimer
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class RetryFlowCoordinator[In, State, Out](
+    parallelism: Long,
+    retryWith: PartialFunction[(Try[Out], State), Option[immutable.Iterable[(In, State)]]],
+    minBackoff: FiniteDuration,
+    maxBackoff: FiniteDuration,
+    randomFactor: Double)
+    extends GraphStage[
+      BidiShape[(In, State), (Try[Out], State), (Try[Out], State, InternalState), (In, State, InternalState)]] {
+
+  private val externalIn = Inlet[(In, State)]("RetryFlow.externalIn")
+  private val externalOut = Outlet[(Try[Out], State)]("RetryFlow.externalOut")
+
+  private val internalIn = Inlet[(Try[Out], State, InternalState)]("RetryFlow.internalIn")
+  private val internalOut = Outlet[(In, State, InternalState)]("RetryFlow.internalOut")
+
+  override val shape
+      : BidiShape[(In, State), (Try[Out], State), (Try[Out], State, InternalState), (In, State, InternalState)] =
+    BidiShape(externalIn, externalOut, internalIn, internalOut)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+    private final val NonoTime = System.nanoTime()
+    private var numElementsInCycle = 0
+    private var queueRetries =
+      scala.collection.immutable.SortedSet.empty[(In, State, InternalState)](Ordering.fromLessThan { (e1, e2) =>
+        if (e1._3.retryDeadline != e2._3.retryDeadline) e1._3.retryDeadline < e2._3.retryDeadline
+        else e1.hashCode < e2.hashCode
+      })
```

Comment:
Something isn't quite right with this `queueRetries`. Theoretically entries could have same `retryDeadline` and same `hashCode` resulting that the `SortedSet` will not hold both entries.

Looks like the Set is used for two things. Check for empy and find the first (with smallest retryDeadLine). Maybe this should be an ordinary `Set` or `List` and finding the smallest can be done on demand with `minBy`?

Also, when are entries removed from the Set? I can see removal in onTimer but it looks like there could be cases where entry is added to queueRetries without scheduleRetryTimer. Not sure I follow the logic here.